psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -720,192 +768,882 @@ void DirectJK::build_JK(std::vector<std::shared_ptr<TwoBodyAOInt>>& ints, std::v
+                        double* K1p = KTp[0L * max_functions_per_shell];
+                        double* K2p = KTp[1L * max_functions_per_shell];
+                        double* K3p = KTp[2L * max_functions_per_shell];
+                        double* K4p = KTp[3L * max_functions_per_shell];
+
+                        double prefactor = 1.0;
+                        if (P == Q) prefactor *= 0.5;
+                        if (R == S) prefactor *= 0.5;
+                        if (P == R && Q == S) prefactor *= 0.5;
+
+                        for (int p = 0; p < shell_P_nfunc; p++) {
+                            for (int q = 0; q < shell_Q_nfunc; q++) {
+                                for (int r = 0; r < shell_R_nfunc; r++) {
+                                    for (int s = 0; s < shell_S_nfunc; s++) {
+
+                                        K1p[(p + shell_P_offset) * nRbasis + r + shell_R_offset] +=
+                                            prefactor * (Dp[q + Qoff][s + Soff]) * (*buffer2);
+                                        K2p[(p + shell_P_offset) * nSbasis + s + shell_S_offset] +=
+                                            prefactor * (Dp[q + Qoff][r + Roff]) * (*buffer2);
+                                        K3p[(q + shell_Q_offset) * nRbasis + r + shell_R_offset] +=
+                                            prefactor * (Dp[p + shell_P_start][s + Soff]) * (*buffer2);
+                                        K4p[(q + shell_Q_offset) * nSbasis + s + shell_S_offset] +=
+                                            prefactor * (Dp[p + shell_P_start][r + Roff]) * (*buffer2);
+
+                                        buffer2++;
+                                    }
+                                }
+                            }
                         }
                     }
+                    touched = true;
                 }
             }
+        }
+
+        // => Master shell quartet loops <= //
+
+        if (!touched) continue;
+
+        // => Stripe out <= //
+
+        for (size_t ind = 0; ind < D.size(); ind++) {
+            double** KTp = KT[thread][ind]->pointer();
+            double** Kp = K[ind]->pointer();
+
+            double* K1p = KTp[0L * max_functions_per_shell];
+            double* K2p = KTp[1L * max_functions_per_shell];
+            double* K3p = KTp[2L * max_functions_per_shell];
+            double* K4p = KTp[3L * max_functions_per_shell];
 
             // > K_PR < //
 
-            for (int P2 = 0; P2 < nPtask; P2++) {
-                for (int R2 = 0; R2 < nRtask; R2++) {
-                    int P = task_shells[P2start + P2];
-                    int R = task_shells[R2start + R2];
-                    int Psize = primary_->shell(P).nfunction();
-                    int Rsize = primary_->shell(R).nfunction();
-                    int Poff = primary_->shell(P).function_index();
+            for (int P = Pstart; P < Pstart + nPshell; P++) {
```

Comment:
I wonder if it would be more efficient to combine the P portions of the K_PR and K_PS loops. That is, have a single loop over P, with the loops over R and S contained within that P loop, but separate from each other, if that makes sense.

Reply:
That is a good point, I think I will do that. Probably minimal performance gain, but wouldn't hurt.